TBRANDS 65 add carousel indicators

JackHowa opened the pull request:
## Ticket

- [TBRANDS-65](https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TBRANDS-65&assignee=5e387d6a1b1d910e5dfd7a9b)

## Description

Add product gallery with thumbnail and dots layout 

## Acceptance Criteria

This should be an additive update to the Carousel component

New options to the Carousel to have indicators either none,dots,thumbnails with none being the default

## Test Steps

1. Checkout branch - `git checkout TBRANDS-65-carousel-thumbnails`
2. Update dependencies - `npm i`
3. Run Storybook `npm run storybook`
4. Look at the dot and thumbnail carousel stories (last) 
5. Ensure that the dots go to the correct slide 
6. Look at the thumbnail one on a smaller screen to see the thumbnail highlight and scroll into view if hidden previously. 

## Dependencies 

- Will implement this in the blocks. If this was merged we could use it in the blocks storybook 
- Feature pack pr https://github.com/WPMedia/arc-themes-feature-pack/pull/351
- Goes with blocks pr https://github.com/WPMedia/arc-themes-blocks/pull/1513

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**


---
notes 

# Testing 

- `npm run storybook`
- Go to the carousel story with indicator dots
- Click around the buttons
- See the carousel story for the thumbnail 
- [x] Todo: Show active styles around the active image 
- [x] Todo: Add tests for thumbnail 
- [x] Add cover affecting transparency of the button based on active and non-active image
- [x] Add scroll into view


---

JackHowa committed:
feat: Add indicator dots and other indicator logic including go to helper

JackHowa commented on src/components/carousel/_index.scss:
```diff
@@ -45,7 +45,7 @@
 			calc(
 				var(--slide-width) -
-				calc(var(--slide-gap) * calc(var(--viewable-slides) - 1) / var(--viewable-slides))
+					calc(var(--slide-gap) * calc(var(--viewable-slides) - 1) / var(--viewable-slides))
```
prettier seemed to contradict stylelint on this

JackHowa replied:
updated

JackHowa committed:
test: Add indicator test for child component with mock

JackHowa committed:
test: Update coverage for child component indicator area dots

JackHowa committed:
refactor: Simplify helper autoplay

JackHowa committed:
refactor: Update the width 1 global

JackHowa committed:
feat: Match variables of designated in color properties

JackHowa committed:
refactor: Use singular dot rather than plural for classname/token

JackHowa committed:
feat: Add thumbnail support for components carousel

matthewroach commented on src/components/carousel/_children/IndicatorArea.jsx:
```diff
@@ -0,0 +1,64 @@
+			Array.from({ length: totalSlideNumber }, (_unusedValue, index) => (
+				<button
+					// todo: translation for slide number
+					label={`Go to slide ${index + 1}`}
```
This should be `aria-label`

```suggestion
					aria-label={`Go to slide ${index + 1}`}
```

matthewroach commented on src/components/carousel/_children/IndicatorArea.jsx:
```diff
@@ -0,0 +1,64 @@
+		{indicatorType === "thumbnails" &&
+			getImages(children).map((child, index) => (
+				<button
+					label={`Go to slide ${index + 1}`}
```
This should be `aria-label`

```suggestion
					aria-label={`Go to slide ${index + 1}`}
```

matthewroach replied:
How are we are handling scrolling on the thumbnail track?

JackHowa replied:
updated

matthewroach commented on src/components/carousel/_children/IndicatorArea.jsx:
```diff
@@ -0,0 +1,64 @@
+
+// get only the images from the nested sub components
+// recursive loop to only find the Image type component
+const getImages = (children) =>
```
I have a concern with this approach, as the images used within the main part of the carousel will not be fully configured to be placed in multiple locations. What are we suggesting for ensuring the images used as thumbnails are sized accordingly?

JackHowa replied:
That's a good concern. I was planning on adding alternate really small sizes to the sizes array to any image that could be used in a thumbnail approach. What did you have in mind as an opportunity here?

JackHowa replied:
Will take in thumbnails as react children for the props

JackHowa replied:
addressed

matthewroach reviewed (commented):
I see it's still in draft and work in progress, but wanted to add some initial feedback from what I can see already. Feel free to ignore if these items are being addressed, or are still to be done

Commit:
Update src/components/carousel/_children/IndicatorArea.jsx

Co-authored-by: Matthew Roach <matthewroach@users.noreply.github.com>

Commit:
Update src/components/carousel/_children/IndicatorArea.jsx

Co-authored-by: Matthew Roach <matthewroach@users.noreply.github.com>

JackHowa commented:
> How are we are handling scrolling on the thumbnail track?

I'm still working on that. As far as I heard from design, we wanted it to be a fluid scroll. I think we talked about having a scrollbar here. 

@matthewroach

Commit:
Merge branch 'arc-themes-release-version-2.0.2' into TBRANDS-65-carousel-thumbnails

JackHowa committed:
feat: Add thumbnail array handling. Refactor dots and separate thumbnail logic

JackHowa committed:
fix: Flip the overflow and ensure that only x is scrollable

JackHowa committed:
fix: Add height to accomodate for scrollbar

JackHowa committed:
feat: Add scroll into view functionality for thumbnails

JackHowa committed:
feat: Update opacity on the thumbnails

JackHowa committed:
ci: Remove contradictory stylelint rule. Add cli command to check in the future

JackHowa committed:
docs: Add explanation of the thumbnails and indicators

JackHowa committed:
docs: Clarify text in docs

JackHowa commented on src/components/carousel/_index.scss:
```diff
@@ -45,7 +45,7 @@
 			calc(
 				var(--slide-width) -
-				calc(var(--slide-gap) * calc(var(--viewable-slides) - 1) / var(--viewable-slides))
+					calc(var(--slide-gap) * calc(var(--viewable-slides) - 1) / var(--viewable-slides))
```
updated so that stylelint should ignore

JackHowa committed:
ci: Ignore word break for styelint

JackHowa commented:
Based on talking in standup, this chunk of work is ready for review. Will work on integrating with the brands block styles

matthewroach commented on .storybook/themes/commerce.scss:
```diff
@@ -334,6 +334,46 @@
+				carousel-indicator-thumbnails-container: (
+					padding: var(--global-spacing-4) var(--global-spacing-3) 0 var(--global-spacing-3),
+					width: 100%,
+					overflow-x: scroll,
```
If we use `auto` this will only place the scroll bar when needed and we avoid having the default browser UI for scrollbar showing at all times

```suggestion
					overflow-x: auto,
```

JackHowa replied:
cool good improvement

matthewroach commented on .storybook/themes/commerce.scss:
```diff
@@ -334,6 +334,46 @@
+				),
+				carousel-indicator-thumbnail-active: (
+					border: var(--global-border-width-1) solid var(--global-black),
+					opacity: 75%,
```
There is no `opacity` in the designs for the active thumbnail which would be this value would be `100%`

JackHowa replied:
oh huh. it must've changed since yesterday

<img width="860" alt="Screen Shot 2022-09-13 at 10 53 17" src="https://user-images.githubusercontent.com/5950956/189948796-8a6e5922-1fd2-44e7-8e4a-80a4422416c8.png">

yep


<img width="916" alt="Screen Shot 2022-09-13 at 10 54 43" src="https://user-images.githubusercontent.com/5950956/189949145-57ab43cf-77a2-495e-b03b-e368d6b9b7ab.png">

JackHowa replied:
oh I see, that previous design was for the icon not the picture

JackHowa replied:
updated -- thanks

matthewroach commented on src/components/carousel/_children/IndicatorArea.jsx:
```diff
@@ -0,0 +1,75 @@
+			<div className={`${BLOCK_CLASS_NAME}__indicator-dots-container`}>
+				{Array.from({ length: totalSlideNumber }, (_unusedValue, index) => (
+					<button
+						// todo: translation for slide number
```
Can we address the todo?

JackHowa replied:
Yeah will do

JackHowa replied:
updated

matthewroach commented on src/components/carousel/index.stories.mdx:
```diff
@@ -681,3 +683,84 @@ const Feature = () => (
+	</Story>
+</Canvas>
+
+** Indicators thumbnails **
```
As the thumbnail indicator has scroll into view logic - should we add a story with a lot more images to be able to see this working?

JackHowa replied:
Yeah, good idea. I suppose I was shrinking the window to see the effect

JackHowa replied:
updated

JackHowa committed:
fix: Remove active opacity on the thumbnail

JackHowa committed:
fix: Set default value for opacity on active

Commit:
Use auto rather than scroll

Co-authored-by: Matthew Roach <matthewroach@users.noreply.github.com>

JackHowa committed:
feat: Add internalization phrase handler for aria label of indicators.

Include fallback phrase like other phrase handlers. Add alphabetization for documentation's sake and readability. via https://github.com/WPMedia/arc-themes-components/pull/142#discussion_r969768730

JackHowa committed:
refactor: Add language fallback and props to default props. Separate the indicator areas.

Addresses https://github.com/WPMedia/arc-themes-components/pull/142#discussion_r969768730

JackHowa committed:
Merge branch 'TBRANDS-65-carousel-thumbnails' of https://github.com/WPMedia/arc-themes-components into TBRANDS-65-carousel-thumbnails

JackHowa committed:
docs: Update to add thumbnail with many indicators based on feedback

via https://github.com/WPMedia/arc-themes-components/pull/142#discussion_r969770941

JackHowa commented:
@matthewroach changes have been addressed. thanks for the feedback

JackHowa committed:
fix: Update aliases for the border style

matthewroach commented on src/components/carousel/index.jsx:
```diff
@@ -419,6 +445,9 @@ Carousel.defaultProps = {
 	},
 	enableAutoplay: false,
+	indicators: "none",
+
```
Empty line - we can remove this

JackHowa replied:
updated

matthewroach commented on src/components/carousel/_children/DotIndicatorArea.jsx:
```diff
@@ -0,0 +1,35 @@
+	/** Function to go to a slide */
+	goToSlide: PropTypes.func.isRequired,
+	/** Internationalization function for handling indicator aria-labels  */
+	goToSlidePhrase: PropTypes.func,
```
Based on the code in line 9 - The prop `goToSlidePhrase` is required - we should ensure this is set on the propTypes

```suggestion
	goToSlidePhrase: PropTypes.func.isRequired,
```

JackHowa replied:
updated -- had to check it locally. thanks

matthewroach reviewed (changes requested):
Looking good - just a couple of minor items

JackHowa committed:
fix: Update proptypes definition to be required based on code

via https://github.com/WPMedia/arc-themes-components/pull/142#discussion_r970639459

JackHowa committed:
style: Remove empty line

via https://github.com/WPMedia/arc-themes-components/pull/142#discussion_r970636783

JackHowa commented:
@matthewroach thanks for the feedback. changes have been addressed 😄

matthewroach commented:
This can be merged ahead of the other PR's as this is an additive change and does not impact existing blocks
